gluonnlp/vocab.py

```diff
@@ -456,3 +460,226 @@ def from_json(json_str):
+
+        Parameters
+        ----------
+        subwordindices : iterable of int
+            Subword indices to look up.
+
+        Returns
+        -------
+        Iterable of str.
+
+        """
+        raise NotImplementedError
+
+    def subwords_to_indices(self, subwords):
+        """Return list of subwordindices associated with subwords.
+
+        Parameters
+        ----------
+        subwords : iterable of str
+            Subwords to replace by indices.
+
+        Returns
+        -------
+        Iterable of int.
+
+        """
+        raise NotImplementedError
+
+
+@register_subword_function
+class ByteSubwords(SubwordFunction):
+    """Map words to a list of bytes.
+
+    Parameters
+    ----------
+    encoding : str, default 'utf-8
+        Encoding to use for obtaining bytes.
+
+    """
+
+    def __init__(self, encoding='utf-8'):
+        self.encoding = encoding
+
+    def __call__(self, words):
+        return [
+            np.frombuffer(word.encode(self.encoding), dtype=np.uint8).astype(
+                np.int_) for word in words
+        ]
+
+    def __len__(self):
+        return 256
+
+    def __repr__(self):
+        return 'ByteSubwords(encoding={})'.format(self.encoding)
+
+    def indices_to_subwords(self, indices):
+        """Return list of subwords associated with subword indices.
+
+        This may raise RuntimeError if the subword function is not invertible.
+
+        Parameters
+        ----------
+        subwordindices : iterable of int
+            Subword indices to look up.
+
+        Returns
+        -------
+        Iterable of str.
+
+        """
+        return indices
+
+    def subwords_to_indices(self, subwords):
+        """Return list of subwordindices associated with subwords.
+
+        Parameters
+        ----------
+        subwords : iterable of str
+            Subwords to replace by indices.
+
+        Returns
+        -------
+        Iterable of int.
+
+        """
+        return subwords
+
+
+@register_subword_function
+class NGramHashes(SubwordFunction):
+    """Map words to a list of hashes in a restricted domain.
+
+    The hash function is the same as in
+    https://github.com/facebookresearch/fastText
+
+    Parameters
+    ----------
+    num_subwords : int
+        Size of target set for the hash function.
+    ngrams : list of int, default [3, 4, 5, 6]
+        n-s for which to hash the ngrams
+
+    """
+
+    def __init__(self, num_subwords, ngrams=(3, 4, 5, 6)):
+        self.num_subwords = num_subwords
+        self.ngrams = ngrams
+
+        # Information for __repr__
+        self.ngrams = ngrams
+
+    @staticmethod
+    def fasttext_hash_asbytes(s, encoding='utf-8'):
+        h = np.uint32(2166136261)
+        s = s.encode(encoding)
+        old_settings = np.seterr(all='ignore')
+        for c in s:
+            h = h ^ np.uint32(c)
+            h = h * np.uint32(16777619)
+        np.seterr(**old_settings)
+        return h
```

Comment:
Would it be slow if we write it in python?

Reply:
Yes, but for training the subword indices for all words should be precomputed. Already the padding and concatenation of the subword indices for the variable length words in a batch takes too much time.